### PR TITLE
National cloud

### PR DESCRIPTION
**Pull request by hforeste**
## Overview

Applens UI and backend UI project has evolved into a state that it wasn't compatible with soveriegn clouds and there also runtime errors for services that were not injected. This change updates the project to be compatible and it also provides a better framework for developers to contribute to the project while thinking of sovereign clouds.

## Related Work Item

Internal DevOps

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)

### This PR:

- Cleanly separate setting up of configurations in `StartupPublicAzure` and `StartupNationalCloud` by calling `AddConfigurations`
- Enable clientCertAuth with Runimehost in all clouds.
  - Add `DiagnosticRole:clientCertEnabled` setting for flexibility
- We've disabled the UI's Adal functionality in non-PublicAzure and replaced with dSTS. This has been the case for a long time but with the UserSettings feature
- Add placeholder API for getuserinfo when user_id is blank
   - UI makes a GET to getuserinfo API with no user_id because Adal doesn't run in non-PublicAzure scenario. This was previously breaking the request flow with a 415 Method Not Found status code.
- Update dSTS auth flow to challenge the request in the case of an unauthenticated user. This is a more industry standard approach as opposed to the redirects that was there previously.
   - Currently when the session cookie expires for dSTS and a user is still in the middle of using detectors. The detectors will immediately fail with no warning as to why. This is because when the UI code is calling the backend, the API challenges the unauthenticated request but the UI code doesn't recognize the redirect to dSTS. We need future bug fix for this but its not blocking the PR.
- Create a base appsettings.json file for national cloud scenarios so as to not repeat settings in individual cloud appsettings.

## How Can This Be Tested? <span>&#128269;</span>

- Local Testing
- Testing dSTS auth flow with SSL enabled
- Verified PublicAzure and non-PublicAzure scenarios are working fine
  - Load detectors
  - Ran applicationcrashes and another detector 

## Checklist <span>&#9989;</span>
- [X] I have looked over the diffs.
- [ ] I have run the linter to catch any errors.
- [X] There are no errors from my changes on this branch.
- [N/A] I have ensured that my changes support accessibility and can be accessed with the keyboard alone
- [X] I have requested review on this PR.

### CODE REVIEW

File: ApplensBackend/Configuration/Startup.cs [hunk 1]
```diff
@@ -119,8 +114,10 @@ public void ConfigureServices(IServiceCollection services)
                 GraphTokenService.Instance.Initialize(Configuration);
             }
 
-            // If we are using runtime host directly
-            DiagnosticClientToken.Instance.Initialize(Configuration);
+            if (!Configuration.GetValue<bool>("DiagnosticRole:clientCertEnabled", true))
+            {
+                DiagnosticClientToken.Instance.Initialize(Configuration);
```
Comment (rekhaswaminathan) [inline ApplensBackend/Configuration/Startup.cs:119]: Should we clean `DiagnosticClientToken` up since it is not needed anymore?

Comment (hforeste) [reply ApplensBackend/Configuration/Startup.cs:119]: I wanted to move everything to DiagnosticClientToken in national clouds first before doing this

File: ApplensBackend/Configuration/StartupNationalCloud.cs [hunk 2]
```diff
@@ -1,24 +1,69 @@
 ﻿using System.IO;
```
Comment (yifguoMSFT) [inline ApplensBackend/Configuration/StartupNationalCloud.cs:1]: now that we have cloud specific app settings, do we still need an extra startup.cs for national cloud?

**Review by yifguoMSFT — APPROVED** [review]

File: ApplensBackend/Configuration/StartupNationalCloud.cs (hunk 2 above)
Comment (hforeste) [reply ApplensBackend/Configuration/StartupNationalCloud.cs:1]: Honestly, we can rely on just one startup.cs file we'd just have to add conditional statements for different scenarios. I prefer to hide implementation details by classes versus conditional logic.

**Review by rekhaswaminathan — APPROVED** [review]

**Pull request merged**